input-profile.js

```diff
@@ -0,0 +1,249 @@
+            this.loadModel(xrInputSource);
+        } else {
+            fetch(this.url)
+                .then((res) => res.json())
+                .then((out) => {
+                    this.ProfileJSON = out;
+                    this._profileJsonLoaded = true;
+                    this.loadModel(xrInputSource);
+                })
+                .catch((error) => {
+                    console.log('couldnot read profile.json ' + this.url);
+                });
+        }
+    }
+
+    loadModel(xrInputSource) {
+        if (this.handedness === xrInputSource.handedness) {
+            this.gamepad = xrInputSource.gamepad;
+            /** load controllerModel **/
+
+            if (this.modelLoaded) return;
+
+            if (this.preSpawnedController !== null) {
+                this.controllerObject = this.preSpawnedController;
+                this.controllerObject.parent = this.object;
+                this.controllerObject.setTranslationLocal([0, 0, 0]);
+                this.getGamepadObjectsFromProfile(this.ProfileJSON, this.controllerObject);
+                this.modelLoaded = true;
+                console.log('Loaded ' + this.handedness + '  controller');
+            } else {
+                const assetPath = this.profile + '/' + this.handedness + '.glb';
+                this.engine.scene
+                    .append(assetPath)
+                    .then((obj) => {
+                        this.controllerObject = obj;
+                        this.components = [];
+                        this.getComponents(obj);
+                        this.controllerObject.parent = this.object;
+                        this.controllerObject.setTranslationLocal([0, 0, 0]);
+                        this.getGamepadObjectsFromProfile(
+                            this.ProfileJSON,
+                            this.controllerObject
+                        );
+                        this.modelLoaded = true;
+                        console.log('Loaded ' + this.handedness + '  controller');
+                    })
+                    .catch(console.error);
+            }
+        }
+    }
+
+    getGamepadObjectsFromProfile(profile, obj) {
+        //initialise components
+        const components = profile.layouts[this.handedness].components;
+        if (!components) return;
+        for (const i in components) {
+            if (components.hasOwnProperty(i)) {
+                const visualResponses = components[i].visualResponses;
+                for (const j in visualResponses) {
+                    if (visualResponses.hasOwnProperty(j)) {
+                        this.getObjectByName(obj, visualResponses[j].valueNodeName);
+                        this.getObjectByName(obj, visualResponses[j].minNodeName);
+                        this.getObjectByName(obj, visualResponses[j].maxNodeName);
+                    }
+                }
+            }
+        }
+    }
+
+    getObjectByName(obj, name) {
```

Comment:
I think you can entirely remove this function, and now use `Object3D.findByName`?
https://wonderlandengine.com/jsapi/object3d/#findByNameRecursive

Reply:
I didnt name the function right , its intended to save the objec to gamepadObjects array , will fix that